core/arch/arm/mm/core_mmu.c

```diff
@@ -1354,18 +1354,30 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
 			 */
 			unsigned int idx;
 
-			assert(*pgt); /* We should have alloced enough */
-
 			/* Virtual addresses must grow */
 			assert(r.va > pg_info->va_base);
 
 			idx = core_mmu_va2idx(dir_info, r.va);
-			pg_info->table = (*pgt)->tbl;
 			pg_info->va_base = core_mmu_idx2va(dir_info, idx);
+
 #ifdef CFG_PAGED_USER_TA
+			/*
+			 * Advance pgt to va_base, note that we may need to
+			 * skip multiple page tables if there are large
+			 * holes in the vm map.
+			 */
+			while ((*pgt)->vabase < pg_info->va_base) {
+				*pgt = SLIST_NEXT(*pgt, link);
+				/* We should have alloced enough */
+				assert(*pgt);
+			}
 			assert((*pgt)->vabase == pg_info->va_base);
-#endif
+			pg_info->table = (*pgt)->tbl;
+#else
+			assert(*pgt); /* We should have alloced enough */
+			pg_info->table = (*pgt)->tbl;
```

Comment:
Is it expected that we do `*pgt = SLIST_NEXT(*pgt, link);` after this line, but we don't in the first branch of the `#ifdef`?

Reply:
Yes, there's no `vabase` in the `pgt` without `CFG_PAGED_USER_TA` so the pgt are altways completely rewritten each time `core_mmu_populate_user_map()` is called. If we'd ever cache the filled in tables even without `CFG_PAGED_USER_TA` we would perhaps need to do something about that.

Reply:
OK, then for commit `core: fix assertion '(*pgt)->vabase == pg_info->va_base'`:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`